File: leetcode/0073_set_matrix_zeroes/solution.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        rows, cols = len(matrix), len(matrix[0])
        first_row_zero = any(val == 0 for val in matrix[0])
        first_col_zero = any(row[0] == 0 for row in matrix)

        for i, row in enumerate(matrix[1:], start=1):
            for j, val in enumerate(row[1:], start=1):
                if val == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0

        for i, row in enumerate(matrix[1:], start=1):
            for j in range(1, cols):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        if first_row_zero:
            for j in range(cols):
                matrix[0][j] = 0

        if first_col_zero:
            for i in range(rows):
                matrix[i][0] = 0

File: leetcode/0073_set_matrix_zeroes/solution.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        bool first_row_zero = false;
        bool first_col_zero = false;


        for (int i = 0; i < cols; i++) {
            if (matrix[0][i] == 0) {first_row_zero = true; break;}
        }

        for (int i = 0; i < rows; i++) {
            if (matrix[i][0] == 0) {first_col_zero = true; break;}
        }

        for (int i = 1; i < rows; ++i) {
            for (int j = 1; j < cols; ++j) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }

        for (int i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (matrix[0][j] == 0 || matrix[i][0] == 0) matrix[i][j] = 0;
            }
        }

        if (first_row_zero) {
            for (int j = 0; j < cols; j++) {
                matrix[0][j] = 0;
            }
        }

        if (first_col_zero) {
            for (int i = 0; i < rows; i++) {
                matrix[i][0] = 0;
            }
        }

    }
};